src/api/channel.rs

```diff
@@ -0,0 +1,649 @@
+    let (mut inner, pool) = self.setup()?;
+
+    inner.limit = Some(frame_limit);
+
+    // TODO: make it user-settable
+    let input_len = self.enc.rdo_lookahead_frames as usize * 2;
+
+    let (send_frame, receive_frame) = bounded(input_len);
+    let (send_packet, receive_packet) = unbounded();
+    let (send_rc_pass2, mut receive_rc_pass2) = unbounded();
+
+    let enc = Arc::new(self.enc);
+
+    let channel = (
+      FrameSender::new(frame_limit, send_frame, enc.clone()),
+      PacketReceiver { receiver: receive_packet, enc },
+    );
+
+    fn feed_pass_data<T: Pixel>(
+      inner: &mut ContextInner<T>, recv: &mut Receiver<RcData>,
+    ) -> Result<(), ()> {
+      while inner.rc_state.twopass_in_frames_needed() > 0
+        && !inner.done_processing()
+      {
+        if let Ok(RcData::Frame(data)) = recv.recv() {
+          inner
+            .rc_state
+            .parse_frame_data_packet(data.as_ref())
+            .unwrap_or_else(|_| todo!("Error reporting"));
+        } else {
+          todo!("Error reporting");
+        }
+      }
+
+      Ok(())
+    }
+
+    pool.spawn(move || {
+      for f in receive_frame.iter() {
+        // info!("frame in {}", inner.frame_count);
+        while !inner.needs_more_fi_lookahead() {
+          feed_pass_data(&mut inner, &mut receive_rc_pass2).unwrap();
+          // needs_more_fi_lookahead() should guard for missing output_frameno
+          // already.
+          //
+          // this call should return either Ok or Err(Encoded)
+          match inner.receive_packet() {
+            Ok(p) => {
+              send_packet.send(p).unwrap();
+            }
+            Err(EncoderStatus::Encoded) => {}
+            Err(EncoderStatus::NotReady) => todo!("Error reporting"),
+            e => unreachable!("Got {:?}", e),
+          };
+        }
+
+        let (frame, params) = f;
+        let _ = inner.send_frame(frame, params); // TODO make sure it cannot fail.
+      }
+
+      inner.limit = Some(inner.frame_count);
+      let _ = inner.send_frame(None, None);
+
+      loop {
+        feed_pass_data(&mut inner, &mut receive_rc_pass2).unwrap();
+        let r = inner.receive_packet();
+        match r {
+          Ok(p) => {
+            // warn!("Sending out {}", p.input_frameno);
+            send_packet.send(p).unwrap();
+          }
+          Err(EncoderStatus::LimitReached) => break,
+          Err(EncoderStatus::Encoded) => {}
+          e => unreachable!("Got {:?}", e),
+        };
+      }
+    });
+
+    Ok((channel, RcDataSender::new(pass_limit, send_rc_pass2)))
+  }
+
+  /// Create a multipass encoder channel
+  ///
+  /// The `PacketReceiver<T>` may block if not enough pass statistics data
+  /// are sent through the `PassDataSender` endpoint
+  ///
+  /// Drop the `FrameSender<T>` endpoint to flush the encoder.
+  /// The last buffer in the PassDataReceiver is the summary of the whole
+  /// encoding process.
+  pub fn new_multipass_channel<T: Pixel>(
```

Comment:
What is the difference between this and the first+second pass channels? Is the idea that this one would be able to run both passes simultaneously? Or simply a simplified API but still running sequentially?

Reply:
This is to run both passes simultaneously, you take from a previous first pass and emit a new first pass.

Potentially we can have  
``` rust
fn new_multipass_channel<T: Pixel>() -> Result<(VideoDataChannel<T>, Option<RcDataReceiver, Option<RcDataSender>), InvalidConfig>
```
that allocates the right combination of channels based on its configuration.

This is one of the open topics.